File: ros2_ws/src/cpp_pubsub/src/const_br.cpp
#include <memory>

#include "geometry_msgs/msg/transform_stamped.hpp"
#include "rclcpp/rclcpp.hpp"
#include "tf2/LinearMath/Quaternion.h"
#include "tf2_ros/static_transform_broadcaster.h"

#include "angles/angles.h"



#define DEPTH_CAMERA_OFFSET_MM_X 0.0f
#define DEPTH_CAMERA_OFFSET_MM_Y 0.0f
#define DEPTH_CAMERA_OFFSET_MM_Z 1.8f  // The depth camera is shifted 1.8mm up in the depth window


class ConstBr : public rclcpp::Node
{
public:

  std::string panda_base_frame = "panda_link0";
  std::string camera_base_frame = "camera_base";
  std::string depth_camera_frame = "depth_camera_link";

  // define transformation values
  // note: TF does translation before rotation
  double dx = 1.22;
  double dy = 0.6;
  double dz = 1.15;
  double rx = deg2rad(183);
  double ry = deg2rad(30);
  double rz = deg2rad(212);

  ConstBr()
  : Node("const_br")
  {
    tf_static_broadcaster_ = std::make_shared<tf2_ros::StaticTransformBroadcaster>(this);

    // Publish static transforms once upon initialization
    this->publishBaseToPandaTf();
    this->publishDepthToBaseTf();
  }


private:

  void publishBaseToPandaTf()
  { 
    // generate message
    geometry_msgs::msg::TransformStamped t;
    t.header.stamp = this->get_clock()->now();
    t.header.frame_id = panda_base_frame;
    t.child_frame_id = camera_base_frame;

    // translations
    t.transform.translation.x = dx;
    t.transform.translation.y = dy;
    t.transform.translation.z = dz;

    // rotations
    tf2::Quaternion q;
    q.setRPY(rx, ry, rz);
    t.transform.rotation.x = q.x();
    t.transform.rotation.y = q.y();
    t.transform.rotation.z = q.z();
    t.transform.rotation.w = q.w();

    tf_static_broadcaster_->sendTransform(t);
  }

  void publishDepthToBaseTf()
  {
    // This is a purely cosmetic transform to make the base model of the URDF look good.
    geometry_msgs::msg::TransformStamped static_transform;

    static_transform.header.stamp = this->get_clock()->now();
    static_transform.header.frame_id = camera_base_frame;
    static_transform.child_frame_id = depth_camera_frame;

    tf2::Vector3 depth_translation = getDepthToBaseTranslationCorrection();
    static_transform.transform.translation.x = depth_translation.x();
    static_transform.transform.translation.y = depth_translation.y();
    static_transform.transform.translation.z = depth_translation.z();

    tf2::Quaternion depth_rotation = getDepthToBaseRotationCorrection();
    static_transform.transform.rotation.x = depth_rotation.x();
    static_transform.transform.rotation.y = depth_rotation.y();
    static_transform.transform.rotation.z = depth_rotation.z();
    static_transform.transform.rotation.w = depth_rotation.w();

    tf_static_broadcaster_->sendTransform(static_transform);
  }

  tf2::Vector3 getDepthToBaseTranslationCorrection()
  {
    // These are purely cosmetic tranformations for the URDF drawing!!
    return tf2::Vector3(DEPTH_CAMERA_OFFSET_MM_X / 1000.0f, DEPTH_CAMERA_OFFSET_MM_Y / 1000.0f,
                        DEPTH_CAMERA_OFFSET_MM_Z / 1000.0f);
  }

  tf2::Quaternion getDepthToBaseRotationCorrection()
  {
    // These are purely cosmetic tranformations for the URDF drawing!!
    tf2::Quaternion ros_camera_rotation;  // ROS camera co-ordinate system requires rotating the entire camera relative to
                                          // camera_base
    tf2::Quaternion depth_rotation;       // K4A has one physical camera that is about 6 degrees downward facing.

    depth_rotation.setEuler(0, angles::from_degrees(-6.0), 0);
    ros_camera_rotation.setEuler(M_PI / -2.0f, M_PI, (M_PI / 2.0f));

    return ros_camera_rotation * depth_rotation;
  }

  double deg2rad(double deg) {
    return (deg / 180) * M_PI;
  }

  std::shared_ptr<tf2_ros::StaticTransformBroadcaster> tf_static_broadcaster_;

};



int main(int argc, char * argv[])
{ 
  // initialize node and spin it
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<ConstBr>());
  rclcpp::shutdown();
  return 0;
}

File: ros2_ws/src/cpp_pubsub/scripts/const_br.py
#!/usr/bin/env python3

import math
import numpy as np

import rclpy
from rclpy.node import Node

from geometry_msgs.msg import TransformStamped
from tf2_ros import TransformBroadcaster


# transformation frames
PANDA_FRAME = "panda_link0"
CAMERA_FRAME = "camera_base"

# translations in the 3 axes
DX = 0.7
DY = 0.0
DZ = 0.0

# rotations around the 3 axes
RX = 0.00
RY = 0.00
RZ = 3.14


####################################################################################
def quaternion_from_euler(ai, aj, ak):
    
    ai /= 2.0
    aj /= 2.0
    ak /= 2.0
    ci = math.cos(ai)
    si = math.sin(ai)
    cj = math.cos(aj)
    sj = math.sin(aj)
    ck = math.cos(ak)
    sk = math.sin(ak)
    cc = ci*ck
    cs = ci*sk
    sc = si*ck
    ss = si*sk

    q = np.empty((4, ))
    q[0] = cj*sc - sj*cs
    q[1] = cj*ss + sj*cc
    q[2] = cj*cs - sj*sc
    q[3] = cj*cc + sj*ss

    return q



####################################################################################
class FramePublisher(Node):

    def __init__(self):

        super().__init__('tf2_camera_frame_publisher')

        self.generate_transformation()

        # Initialize the transform broadcaster
        self.tf_broadcaster = TransformBroadcaster(self)
        self.timer_period = 0.1   # seconds
        self.timer = self.create_timer(self.timer_period, self.broadcast)


    def broadcast(self):
        
        # update time stamp
        self.trans.header.stamp = self.get_clock().now().to_msg()

        # Send the transformation
        print("sending transformation now!")
        self.tf_broadcaster.sendTransform(self.trans)
        

    def generate_transformation(self):

        t = TransformStamped()

        # initialize tf message
        t.header.stamp = self.get_clock().now().to_msg()
        t.header.frame_id = PANDA_FRAME
        t.child_frame_id = CAMERA_FRAME

        # translations
        t.transform.translation.x = DX
        t.transform.translation.y = DY
        t.transform.translation.z = DZ

        # rotations
        q = quaternion_from_euler(RX, RY, RZ)
        t.transform.rotation.x = 0.0
        t.transform.rotation.y = 0.0
        t.transform.rotation.z = 1.0
        t.transform.rotation.w = 0.0

        self.trans = t

        



####################################################################################
def main():

    rclpy.init()
    node = FramePublisher()

    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        print("interrupted!")
        pass

    rclpy.shutdown()


if __name__ == '__main__':
    main()